src/rust/src/lib.rs

```diff
@@ -2,7 +2,86 @@
 // 2.0, and the BSD License. See the LICENSE file in the root of this repository
 // for complete details.
 
+/// Returns the value of the input with the most-significant-bit copied to all
+/// of the bits.
+fn duplicate_msb_to_all(a: u8) -> u8 {
+    0u8.wrapping_sub(a >> 7)
+}
+
+/// This returns 0xFF if a < b else 0x00, but does so in a constant time
+/// fashion.
+fn constant_time_lt(a: u8, b: u8) -> u8 {
+    // Derived from:
+    // https://github.com/openssl/openssl/blob/OpenSSL_1_1_1i/include/internal/constant_time.h#L120
+    duplicate_msb_to_all(a ^ ((a ^ b) | (a.wrapping_sub(b) ^ b)))
+}
+
+#[pyo3::prelude::pyfunction]
+fn check_pkcs7_padding(data: &[u8], block_len: u8) -> bool {
+    let mut mismatch = 0;
+    let pad_size = data[(block_len - 1) as usize];
+    for i in 0..block_len {
+        let mask = constant_time_lt(i, pad_size);
+        let b = data[(block_len - 1 - i) as usize] as u8;
+        mismatch |= mask & (pad_size ^ b);
+    }
```

Comment:
The assignment of `pad_size` will panic if `data` is empty or `block_len` is 0. I'd make that explicit with an 
```rust
assert!(data.len() >= usize::from(block_len) && block_len > 0);
```
as first line. Usually that also helps the optimizer avoid bounds checks but it's not enough in this case.

If you rewrite the loop the rusty way using iterators, the bounds checks disappear and the code gets vectorized:
```rust
// Is `block_len` some C interface remnant? Slices know their length,
// so if `block_len == data.len()` we could skip this.
let data = &data[..usize::from(block_len)];
// This unwrap will always succeed because above we asserted that data.len() > 0.
let pad_size = *data.last().unwrap();
// Usually I'd use `data.iter().enumerate()` here, but then `i` would be u32
// and need a cast to u8 for the call to `constant_time_lt. At least on the
// type level that would be a lossy cast, which I'd rather avoid.
//
// I did consider reversing the range with `rev()` instead of doing it manually
// with `block_len - 1 - i`, but the code looses the vectorizations then.
for (i, b) in (0..block_len).zip(data.iter()) {
    let mask = constant_time_lt(block_len - 1 - i, pad_size);
    mismatch |= mask & (pad_size ^ b);
}
```

PS: Not a crypto person, so take this with the appropriate amount of salt.